### PR TITLE
Support plugin/flags.vim file sourced on first flag access

### LINKED ISSUE #189: Deprecate instant/flags.vim
Maktaba's `instant/` mechanism for executing files immediately has a few drawbacks:
  * Costs overhead on early startup
  * Doesn't interact well with various plugin managers
  * Encourages overuse

It's especially significant if most plugins have an instant/flags.vim file.

We should find another convention for defining flags that doesn't depend on loading them instantly. It could work something like:
  * Flags are defined in whatever standard location for each plugin
  * Maktaba queues up Flag.Set operations until the flag is initialized or read.
  * Flag.Get and any other attempt to access a flag's value will trigger flag initialization.
  * At plugin/ time, all remaining flags are initialized so callbacks and errors trigger.

There'd be some migration pain, and it may not be worth making changes like if there's no measurable performance benefit.

### PR DESCRIPTION
**Pull request by dbarnett**
Partially fixes #189 (allowing plugins to start migrating to plugin/flags.vim but not discouraging the instant/ equivalent yet).

### CODE REVIEW

File: autoload/maktaba/plugin.vim [hunk 1]
```diff
@@ -240,10 +242,10 @@ function! maktaba#plugin#Enter(file) abort
     return [l:plugin, 0]
   endif
 
-  if l:filedir !=# 'autoload'
+  if l:filedir !=# 'autoload' && has_key(l:plugin.flags, l:filedir)
```
Comment (dbarnett) [inline autoload/maktaba/plugin.vim:245]: This change avoids triggering flags.vim to load where it's not needed and potentially wasteful/unsafe. Calling Flag could even try to load recursively if this maktaba#plugin#Enter was called from flags.vim.

**Comment by dbarnett** [comment]
FWIW, this could add a bit of overhead trying to source a plugin/flags.vim file, but only in uncommon corner cases because most current plugins that will have any flags configured will already have `PLUGIN._loaded_flags` set via instant/flags.vim by the time that anything else would try to resolve flags.

Also this approach is a little simpler than what I described in #189 since instead of trying to be clever and queue up flags operations it just tries to ensure flags.vim is sourced before any flags operations are performed to pick up default values. This shouldn't preclude later improvements to do the clever queuing if we're able to get all the intricacies right.

**Review by malcolmr — APPROVED** [review]
If I understand this correctly:

- This is a no-op for any plugin with an `instant/flags.vim`; everything proceeds as normal.
- For any plugin without `instant/flags.vim`, this will trigger an extra load attempt if a caller (not Maktaba) tries to look up its flags (e.g. by using Glaive). But it's unlikely that there are many/any Maktaba plugins that both don't have any flags at all and have a caller that tries to look at their flags.
- This does assume that flags are only defined in `instant/flags.vim` specifically (previously they could have been in _any_ `instant/` file); I suspect that's a reasonable assumption though.

File: autoload/maktaba/plugin.vim [hunk 2]
```diff
@@ -180,6 +180,26 @@ function! s:FileHandleToFlagHandle(handle) abort
...
 
+" Ensures {plugin}'s plugin/flags.vim file has been sourced (if it has one and
+" has not already had an instant/flags.vim file sourced).
+function! s:EnsureFlagsLoaded(plugin) abort
+  if a:plugin._loaded_flags
+    return
+  endif
+  try
+    call a:plugin.Load('flags')
+    if !a:plugin._loaded_flags
```
[the remark this answers is not in the document]
Comment (dbarnett) [reply autoload/maktaba/plugin.vim:193]: Added.

File: autoload/maktaba/plugin.vim [hunk 3]
```diff
@@ -225,6 +245,8 @@ function! maktaba#plugin#Enter(file) abort
   let [l:plugindir, l:filedir, l:handle] = s:SplitEnteredFile(a:file)
   let l:plugin = maktaba#plugin#GetOrInstall(l:plugindir)
   let l:controller = l:plugin._entered[l:filedir]
+  let l:is_flags_file = (l:filedir ==# 'plugin' || l:filedir ==# 'instant')
+      \ && l:handle ==# 'flags^'
```
[the remark this answers is not in the document]
Comment (dbarnett) [reply autoload/maktaba/plugin.vim:253]: Added a comment. It's the weird "handle" syntax from SplitEnteredFile.

File: autoload/maktaba/plugin.vim [hunk 4]
```diff
@@ -240,7 +262,7 @@ function! maktaba#plugin#Enter(file) abort
     return [l:plugin, 0]
   endif
 
-  if l:filedir !=# 'autoload'
+  if l:filedir !=# 'autoload' && !l:is_flags_file
```
[the remark this answers is not in the document]
Comment (dbarnett) [reply autoload/maktaba/plugin.vim:274]: Sure, added some comments to clarify all that.

File: autoload/maktaba/plugin.vim [hunk 5]
```diff
@@ -757,6 +783,8 @@ endfunction
 "   maktaba#plugin#Get('myplugin').Flag('foo')
 "   maktaba#plugin#Get('myplugin').flags.foo.Get()
 " <
+" except that the second version will not trigger flags to load if they haven't
+" been loaded already.
```
[the remark this answers is not in the document]
Comment (dbarnett) [reply autoload/maktaba/plugin.vim:796]: Right, I want to go a little further to discourage accessing `.flags`, possibly rename it to `._flags` and provide a `.Flags()` lookup method instead, but I don't want to get too verbose with usage contracts just yet in all the files I'd need to touch for that. The helpers ought to fully encapsulate having the flags resolved before letting you Set/Get them and I can't think of any reason to support poking at them partially initialized.

File: autoload/maktaba/plugin.vim [hunk 6]
```diff
@@ -506,10 +541,10 @@ function! s:CreatePluginObject(name, location, settings) abort
...
   if isdirectory(maktaba#path#Join([l:plugin.location, 'plugin']))
-    call l:plugin.Flag('plugin', {})
+    let l:plugin.flags.plugin = maktaba#flags#Create('plugin', {})
   endif
   if isdirectory(maktaba#path#Join([l:plugin.location, 'instant']))
-    call l:plugin.Flag('instant', {})
+    let l:plugin.flags.instant = maktaba#flags#Create('instant', {})
   endif
 
   " Load flags file first.
```
Comment (dbarnett) [inline autoload/maktaba/plugin.vim:550]: Actually I noticed this ApplySettings case probably also needs to EnsureFlagsLoaded first. I'm going to try to get some vroom coverage to repro an issue there along with adding the call to fix it.

File: autoload/maktaba/plugin.vim [hunk 7]
```diff
@@ -757,6 +783,8 @@ endfunction
 "   maktaba#plugin#Get('myplugin').Flag('foo')
 "   maktaba#plugin#Get('myplugin').flags.foo.Get()
 " <
+" except that the second version will not trigger flags to load if they haven't
```
[the remark this answers is not in the document]
Comment (dbarnett) [reply autoload/maktaba/plugin.vim]: Sure, updated.

[the remark this answers is not in the document]
Comment (dbarnett) [reply autoload/maktaba/plugin.vim]: In a sense yes, but to actually break anything you'd need to update the plugins to move from instant/flags.vim to plugin/flags.vim.

File: autoload/maktaba/plugin.vim [hunk 8]
```diff
@@ -785,13 +813,16 @@ endfunction
 "   maktaba#plugin#Get('myplugin').Flag('foo', 'bar')
 "   maktaba#plugin#Get('myplugin').flags.foo.Set('bar')
 " <
+" except that the second version will not trigger flags to load if they haven't
```
[the remark this answers is not in the document]
Comment (dbarnett) [reply autoload/maktaba/plugin.vim]: Updated that example to not use `\.flags` so it doesn't need to call out that distinction.

File: autoload/maktaba/plugin.vim (hunk 6 above)
Comment (dbarnett) [reply autoload/maktaba/plugin.vim:551]: Alright, pushed 0382ab2 to add the missing corner case handling (in Setting.Apply) and test coverage.

**Comment by dbarnett** [comment]
Ready for another look after various updates in 0382ab2 to add test coverage and fix ensuring flags file loaded in two cases I'd neglected (`Plugin.HasFlag` and `Setting.Apply`).

**Review by malcolmr — APPROVED** [review]
LGTM.

Nice, thanks!
